Introduce new prune parameter into diff command

ardaguclu opened the pull request:
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
This PR introduces new prune and it's dependent parameters to simulate
`kubectl apply --prune` command. It works exactly the same way how
`kubectl apply --prune` works. Thanks to that, users can run this command 
to find out which resources will be pruned.

#### Which issue(s) this PR fixes:
Fixes #105043

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Added prune flag into diff command to simulate `apply --prune` 
```

ardaguclu committed:
Introduce new prune parameter into diff command

This PR introduces new prune and it's dependent parameters to simulate
`kubectl apply --prune` command.

lauchokyip commented:
/ok-to-test

lauchokyip commented:
Hi @ardaguclu , thanks for the PR, could you add some unit test cases ? for `diff.go` and `prune.go`?

ardaguclu commented:
> Hi @ardaguclu , thanks for the PR, could you add some unit test cases ? for `diff.go` and `prune.go`?

Thanks for review @lauchokyip, I'll add unit tests.

lauchokyip commented:
@ardaguclu Also made sure to add a release note! It definitely introduced changes to the users for this PR

ardaguclu committed:
Add new unit tests for diff/prune

ardaguclu commented:
@lauchokyip I added all possible unit tests. Could you PTAL again?, thanks.

ardaguclu commented:
@apelisse I applied the idea you mentioned in here https://github.com/kubernetes/kubernetes/issues/105043#issuecomment-920149932

Could you please take a look?, thank you.

lauchokyip commented:
> @lauchokyip I added all possible unit tests. Could you PTAL again?, thanks.

I am a bit busy right now, can you post it on slack so other sig-cli members can take a look?

apelisse commented:
cc @seans3

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/prune.go:
```diff
@@ -0,0 +1,236 @@
+/*
```
This file mostly looks like copy-pasted functions from somewhere else? I would have expected the logic to be shared rather than repeated, can we re-use the same piece of code that apply does?

ardaguclu replied:
Thanks for review @apelisse. Yes, you are right. This file is mostly copy-pasted from `apply/prune.go`. If you support the idea moving `apply/prune.go` to some common place, thanks to that unit tests I wrote also cover both of them.

apelisse replied:
Should this live in `cmd/diff`?

apelisse replied:
OK that seems a little different from what we do in the general prune case.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go:
```diff
@@ -107,15 +110,25 @@ type DiffOptions struct {
 	FieldManager    string
 	ForceConflicts  bool
 
-	Selector         string
```
This object already had too many fields, I think it's way too many now, can we refactor this instead?

ardaguclu replied:
Good idea, I'll reorganize structs in here.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go:
```diff
@@ -162,13 +180,26 @@ func NewCmdDiff(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.C
 	cmd.Flags().StringVarP(&options.Selector, "selector", "l", options.Selector, "Selector (label query) to filter on, supports '=', '==', and '!='.(e.g. -l key1=value1,key2=value2)")
+	cmd.Flags().StringArrayVar(&options.PruneWhitelist, "prune-whitelist", options.PruneWhitelist, "Overwrite the default whitelist with <group/version/kind> for --prune")
+	cmd.Flags().BoolVar(&options.Prune, "prune", options.Prune, "Automatically diff for possibly will be deleted resource objects, Should be used with either -l or --all.")
+	cmd.Flags().BoolVar(&options.All, "all", options.All, "Select all resources in the namespace of the specified resource types.")
```
I don't understand how this would work?

ardaguclu replied:
It works exactly the same way how `apply -prune` works. User can `diff --prune --all` or `diff --prune -l label`.

apelisse replied:
That parameter's name is terrible :-)

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go:
```diff
@@ -107,15 +110,25 @@ type DiffOptions struct {
+	PruneResources    []pruneResource
+	VisitedUids       sets.String
+	VisitedNamespaces sets.String
+	ToPrinter         func(string) (printers.ResourcePrinter, error)
+	PrintFlags        *genericclioptions.PrintFlags
```
Why do we need these? How are you planning on showing what was deleted? By showing "deleted" in the output?

ardaguclu replied:
Before showing the differences in file, if `--prune` flag is passed, it'll show;
```
namespace/test pruned
deployment/test pruned
...
```

apelisse replied:
Yeah, that's highly inconsistent with how diff works. If you want to see these, I would suggest you run `kubectl apply --dry-run=client`. The output of kubectl diff should be something that can be diffed. In that specific instance, I would expect the current object to be put in the "old" folder, and no object to be put in the "new" folder, which will force `diff -RN` to show the object as removed.

ardaguclu commented:
@apelisse thank for review and guidance. 

I applied all the changes we discussed. Now prune results are also consistent with usual diff results. And honestly, PR is much nicer than previous version :).

ardaguclu committed:
Update vendor for util/prune

ardaguclu committed:
Add boilerplate header into util/prune

ardaguclu commented:
/test pull-kubernetes-unit

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go:
```diff
@@ -116,6 +119,9 @@ type DiffOptions struct {
 	EnforceNamespace bool
 	Builder          *resource.Builder
 	Diff             *DiffProgram
+	Prune            bool
```
That's a lot of things to track just for pruning, can we move `PruneWhiteList` inside `pruner`? And so that a `nil` `pruner` does nothing (to remove the boolean)

apelisse replied:
I'm seeing below how these are actually command line flags, I'm curious if you have to use the same `options.Flag` mechanism for these if they don't fit well in these options, i.e. use `StringArray` rather than `StringArrayVar` maybe.

apelisse replied:
Can we do the same thing you did for `WhiteList` to this `Prune` field?

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go:
```diff
@@ -131,6 +137,10 @@ func NewDiffOptions(ioStreams genericclioptions.IOStreams) *DiffOptions {
 			Exec:      exec.New(),
 			IOStreams: ioStreams,
 		},
+		pruner: &pruner{
```
Maybe make a constructor for the pruner? That would avoid it from being mis-used

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go:
```diff
@@ -162,6 +172,8 @@ func NewCmdDiff(f cmdutil.Factory, streams genericclioptions.IOStreams) *cobra.C
 	usage := "contains the configuration to diff"
 	cmd.Flags().StringVarP(&options.Selector, "selector", "l", options.Selector, "Selector (label query) to filter on, supports '=', '==', and '!='.(e.g. -l key1=value1,key2=value2)")
+	cmd.Flags().StringArrayVar(&options.PruneWhitelist, "prune-whitelist", options.PruneWhitelist, "Overwrite the default whitelist with <group/version/kind> for --prune")
+	cmd.Flags().BoolVar(&options.Prune, "prune", options.Prune, "Automatically diff for possibly will be deleted resource objects, Should be used with either -l or --prune-all.")
```
Can we re-work the description here?

Maybe:
> Include resources that would be deleted by pruning. Should be used with either -l or --prune-all.

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/prune.go:
```diff
@@ -0,0 +1,112 @@
+}
+
+func (p *pruner) GetObjectName(obj runtime.Object) string {
+	// Not compare anything, it is safe to assign random
```
No it's not, the name will give the filename which helps figuring out what resource it represents. Why can't we use the actual name?

ardaguclu replied:
Like you said, naming is important to find the correct mapping of resources in LIVED&MERGED directories. However, in pruning case, it is ensured that resource will be pruned and no need to find comparable file in MERGED.  In any case, this resource will be shown with lots of minuses.

If I'm missing something and it is also mandatory for pruning, this will bring a couple of new fields to generate this naming exactly like other diff; https://github.com/kubernetes/kubernetes/blob/ed40a89f2e16dcefdf57f5f3b24f6f3788d7fc5b/staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go#L399

apelisse replied:
The name is still displayed in the diff and can be used as a hint of what object it is, I don't get why you can't get the name given the `obj` here?

apelisse commented on staging/src/k8s.io/kubectl/pkg/util/prune/prune.go:
```diff
@@ -0,0 +1,105 @@
+func ParseResources(mapper meta.RESTMapper, gvks []string) ([]Resource, error) {
+	pruneResources := []Resource{}
+	for _, groupVersionKind := range gvks {
+		gvk := strings.Split(groupVersionKind, "/")
+		if len(gvk) != 3 {
+			return nil, fmt.Errorf("invalid GroupVersionKind format: %v, please follow <group/version/kind>", groupVersionKind)
+		}
+
+		if gvk[0] == "core" {
+			gvk[0] = ""
+		}
```
I know you're just copying the code, and you probably shouldn't change this, but if you're interested, this code already exists somewhere else in a way that is reusable ;-)

ardaguclu replied:
I think, I'm no more just a copy-paster :) and agree that since this file is being used for actual `apply --prune`, probably we should not change this. But I'm curious of the place where is reusable code exist, maybe we can use.

ardaguclu committed:
Add pruner constructor, tweak DiffOptions

ardaguclu commented:
/test pull-kubernetes-node-e2e-containerd
/test pull-kubernetes-unit

ardaguclu commented:
/test pull-kubernetes-node-e2e-containerd

apelisse commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go:
```diff
@@ -634,6 +641,19 @@ func (o *DiffOptions) Complete(f cmdutil.Factory, cmd *cobra.Command) error {
 		return err
 	}
 
+	if o.Prune {
```
I'm guessing here you would only create the pruner if the prune flag is set to true, and then you can use the `nil` condition on the pruner as a way to know if it's enabled or not (and/or do a no-op on a nil pruner).

apelisse reviewed (commented):
Let me try to see how this works in practice :-)

apelisse commented:
@seans3, would you mind taking a look at the pruning code? Thanks!

ardaguclu committed:
Get object name correctly for pruned resources

apelisse commented:
Thanks for solving all of that! Do you know how we can improve the tests for this? My confidence that this is all working is fairly low because we don't have great tests.

ardaguclu commented:
@apelisse, I tried to implement unit tests to increase quality for diff command. However, since it executes totally different program, it seems not easy. I'd recommend to add more unit tests for diff command in a different PR, because it will not only test just prune but also all diff commands(by mocking the real program finding the diffs between files).

apelisse commented:
We really shouldn't have to test the output of the diff program. Testing that the files are where they are should be good enough.

Unfortunately, it's hard to simulate any of this easily since mocking the pruning isn't easy right now.

ardaguclu committed:
Add prune test into test/cmd/diff.sh

eddiezane reviewed (commented):
Thanks for the PR!

I really like how Brian created `OverrideOptions` as a mixin/module in [this PR](https://github.com/kubernetes/kubernetes/pull/105140). I wonder if that would be a fit to adapt here?

ardaguclu committed:
Refactor diff/prune

This PR does some refactors for diff/prune;

- GetRESTMappings takes value array instead reference
- Move getObjectName into diff instead prune
- License, etc. changes

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/util/prune/prune.go:
```diff
@@ -0,0 +1,105 @@
+func GetRESTMappings(mapper meta.RESTMapper, pruneResources *[]Resource) (namespaced, nonNamespaced []*meta.RESTMapping, err error) {
+	if len(*pruneResources) == 0 {
+		// default allowlist
+		*pruneResources = []Resource{
```
Thanks @eddiezane good catch

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/diff.go:
```diff
@@ -116,6 +118,7 @@ type DiffOptions struct {
 	EnforceNamespace bool
 	Builder          *resource.Builder
 	Diff             *DiffProgram
+	pruner           *pruner
```
This pruner is customized according to diff command usage. No other part of code will need this pruner and I think, it is better not export it.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/prune.go:
```diff
@@ -0,0 +1,151 @@
+	return pobjs, nil
+}
+
+func (p *pruner) GetObjectName(obj runtime.Object) (string, error) {
```
Thanks @brianpursley, you are right, I moved it back to diff.go.

ardaguclu commented on staging/src/k8s.io/kubectl/pkg/cmd/diff/prune.go:
```diff
@@ -0,0 +1,151 @@
+	}
+}
+
+func (p *pruner) pruneAll() ([]runtime.Object, error) {
```
That is a good idea. However, apply pruning works the way whenever it gets error, it returns. On the other hand, diff prune works different, it ignores the errors and walks through all objects to show all possible prunes. 

Although, they are doing the same thing, functionality slightly differs, since apply/prune is transactional and diff/prune is just guidance.

ardaguclu commented:
/retest

pacoxu commented:
/lgtm
/assign soltysh dims

dims commented:
the vendor changes looks good

/approve

ardaguclu commented:
/retest

ardaguclu commented:
/test pull-kubernetes-unit

ardaguclu commented:
/test pull-kubernetes-e2e-kind-ipv6
